### PR TITLE
Add possibility for a `CopyToHost::postCopy()` operation

### LINKED ISSUE #989: Add `postCopy()` option to `CopyToHost` and `CopyToDevice` class templates
See https://github.com/cms-sw/cmssw/issues/45708#issuecomment-2294166204

Done in

- [ ] master: https://github.com/cms-sw/cmssw/pull/45801
- [ ] 14_0_X:

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

Following https://github.com/cms-sw/cmssw/issues/45708#issuecomment-2294166204 this PR adds the possibility for a `CopyToHost<T>::postCopy()` function, that is called by the implicit data product device-to-host copy operation after the copy has finished (but only if the `postCopy()` function is defined). This facility allows data products that need to be updated after a `memcpy()` (e.g. because they have pointers to itself) to be used without blocking synchronization calls in `CopyToHost::copyAsync()`. I expect (hope) we will ever have only at most few such data products.

The first commit has a C++17 SFINAE -based solution for checking if the `CopyToHost<T>::postCopy()` exists (that can be backported to 14_0_X), and the second commit replaces that with C++20 concepts-based solution.

Resolves https://github.com/cms-sw/framework-team/issues/989

#### PR validation:

Added unit test runs on CPU-only and NVIDIA GPU nodes.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

~~To be backported to 14_1_X, and to 14_0_X (first commit only).~~ Following https://github.com/cms-sw/cmssw/issues/45708#issuecomment-2324539478 not to be backported.

### CODE REVIEW

**Comment by makortel** [comment]
enable gpu

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
type -changes-dataformats

The added data format classes are transient (and part of testing suite).

File: DataFormats/PortableTestObjects/interface/TestProductWithPtr.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#define DataFormats_PortableTest_interface_TestProductWithPtr_h
+
+#include "DataFormats/Portable/interface/PortableCollection.h"
+#include "DataFormats/SoATemplate/interface/SoACommon.h"
+#include "DataFormats/SoATemplate/interface/SoALayout.h"
+#include "DataFormats/SoATemplate/interface/SoAView.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/CopyToHost.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/CopyToDevice.h"
+
+#include "alpaka/alpaka.hpp"
```
Comment (fwyzard) [inline DataFormats/PortableTestObjects/interface/TestProductWithPtr.h:11]:
```suggestion
#include <alpaka/alpaka.hpp>
```

Comment (makortel) [reply DataFormats/PortableTestObjects/interface/TestProductWithPtr.h]: Thanks, done

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
+heterogeneous

@makortel I'm fine with the changes and with this approach.

I'm wondering if there could be a more self-contained approach, where the data structure could define a `postCopy()` or `finalize()` method that is automatically called, instead of adding the static `postCopy()` function in `CopyToHost`.
But that kind of clashes with the fact that SoAs don't generally have methods :-/

**Comment by mandrenguyen** [comment]
+1

**Comment by makortel** [comment]
> I'm wondering if there could be a more self-contained approach, where the data structure could define a `postCopy()` or `finalize()` method that is automatically called, instead of adding the static `postCopy()` function in `CopyToHost`.
> But that kind of clashes with the fact that SoAs don't generally have methods :-/

The thing next closest to a method would be a standalone function. Just to remind, with `CopyToHost` I started with just function overloads, but rejected that approach when it became evident that inheritance would be a tolerated pattern in crafting the data format classes (because a function `copyToHost(T const&)` would bind fine for all types `D` that inherit from `T`, and I wanted those cases to result a compilation error instead because the function must be different for `D`). Therefore I did not consider standalone functions for this case.

Within the class template specialization approach I considered making a separate (say `PostCopy<T>`) class template, because I felt the operation would be a property of `T` rather than of "copying `T` to host". But then I realized the operation needs to be different for copy-to-host and copy-to-device (at minimum copy-to-device needs to be a kernel call), and furthermore for copy-to-device the modification kernel (or `memset()`) can be launched from the `CopyToDevice<T>::copyAsync()` directly (or in case of a kernel probably via a host-side function called from there).

Given that the `postCopy()` is, in the end, tied to the `CopyToHost` operation, and that I hope the `postCopy()` won't be needed widely, I opted in adding it to the `CopyToHost` class template specialization. At least the two parts of the copy operation end up close together in the code.

These decisions can, of course, be revisited if my assumptions above turn out to be wrong.

**Pull request merged**